File: src/services/rotations/functions/KAR_ROT_complete_performance.ts
import { APIGatewayProxyEvent, Handler, APIGatewayProxyResult } from 'aws-lambda';
import * as AWS from 'aws-sdk';
import { middify, formatJSONResponse } from 'lambda-helpers';
import AddPerformance from '../dtos/addPerformanceDto';
import PerformanceCompleted from '../events/PerformanceCompleted';

export const handler: Handler = middify(
  async (event: APIGatewayProxyEvent & AddPerformance): Promise<APIGatewayProxyResult> => {
    const { singerId, songId, songTitle, artist } = event.body;
    const performanceCompleted: PerformanceCompleted = {
      singerId,
      songId,
      songTitle,
      artist,
      performanceDate: new Date().toISOString(),
    };

    console.log(`PerformanceCompleted: ${JSON.stringify(performanceCompleted)}`);

    const params = {
      Entries: [
        {
          Detail: JSON.stringify(performanceCompleted),
          DetailType: process.env.PERFORMANCE_COMPLETED_DETAIL_TYPE,
          Source: process.env.PERFORMANCE_COMPLETED_SOURCE,
          EventBusName: process.env.KARAOKE_BUS,
        },
      ],
    };

    console.log(`Params: ${JSON.stringify(params)}`);

    try {
      const eventBridge = new AWS.EventBridge();
      const result = await eventBridge
        .putEvents(params, (err, data) => {
          if (err) console.log(err, err.stack);
          else console.log(data);
        })
        .promise();

      return formatJSONResponse(201, result);
    } catch (err) {
      return formatJSONResponse(500, err);
    }
  }
);

File: src/services/rotations/KAR_ROT_complete_performance.py
import os
from datetime import datetime
import json
import boto3

def handler(event, context):
    performance_data = json.loads(event['body'])

    performance_completed = {
        'singer_id': performance_data['singer_id'],
        'song_id': performance_data['song_id'],
        'song_title': performance_data['song_title'],
        'artist': performance_data['artist'],
        'performance_date': str(datetime.utcnow())
    }

    entries = [
        {
            'Detail': json.dumps(performance_completed),
            'DetailType': os.environ['PERFORMANCE_COMPLETED_DETAIL_TYPE'],
            'Source': os.environ['PERFORMANCE_COMPLETED_SOURCE'],
            'EventBusName': os.environ['KARAOKE_BUS']
        }
    ]

    client = boto3.client('events')

    response = {}
    response['headers'] = {}
    response['headers']['Content-Type'] = 'application/json'
    try:
        result = client.put_events(Entries=entries)
    except Exception as e:
        print(e)
        response['statusCode'] = 500
        response['body'] = str(e)
    else:
        response['statusCode'] = 200
        response['body'] = json.dumps(result)

    return response
